packages/app/src/cli/services/build/extension.ts

```diff
@@ -124,10 +126,17 @@ export async function buildFunctionExtension(
   extension: ExtensionInstance,
   options: BuildFunctionExtensionOptions,
 ): Promise<void> {
-  if (extension.isJavaScript) {
-    return runCommandOrBuildJSFunction(extension, options)
-  } else {
-    return buildOtherFunction(extension, options)
+  const lockfilePath = joinPath(extension.directory, '.build-lock')
+  const releaseLock = await lockfile.lock(extension.directory, {retries: 10, lockfilePath})
```

Comment:
Also curious if this has any backoff or if it just tries 10 times in a row

Reply:
>What happens if we fail to acquire the lock after 10 retries? Is it an exception?

Yes, it should throw an exception.

>Also curious if this has any backoff or if it just tries 10 times in a row

Based on the code, if you supply an integer, that's passed into the `retry` library. By default, it looks like it's an [exponential backoff](https://github.com/tim-kos/node-retry/blob/master/lib/retry.js#L54).

Reply:
Oops, totally missed your comment @otheriault! GH didn't show it when I typed mine 😓

Without watching, there's nothing else for the process to do (i.e. if we run it without watch, it'd run once and exit). Maybe we should print an error message like you suggested and exit?

Reply:
I also envision a future iteration of this command where we allow developers to press keys to perform actions like in `dev`. For example, I'm thinking of actions like choosing a new run, manually re-running, enabling/disabling watch mode. I think in that future, it makes more sense for us to print a warning and not exit.

Reply:
Okay. So should we print the error, and then let the process continue? It will continue to try to watch (which may or may not work next time).

How should we frame the error message? This error is at the function build level (and it reused whenever we build a function), so the error message could reflect that.

Alternatively, we could bubble the error up and try to catch it at a higher level, but because [this](https://github.com/Shopify/cli/blob/main/packages/app/src/cli/services/dev/extension/bundler.ts#L199-L204) (which is used by `dev`) catches the error and outputs a draft deploy message, I think we'd have to do some refactoring so that the error is handled appropriately in both cases.

Reply:
Spoke offline with @nickwesselman and we're going to try to output an error, but continue watching the files. This involves a little refactoring to make sure the appropriate error is outputted in dev vs. replay.